balancer/proposals/balancer.md

```diff
@@ -0,0 +1,182 @@
+These domains may include:
+
+* Cloud provider zones inside a single region, to ensure that the application is still up and running, even if one of the zones has issues.
+* Different types of Kubernetes nodes. These may involve nodes that are spot/preemptible, or of different machine families. 
+
+A single Kuberentes deployment may either leave the placement entirely up to the scheduler 
+(most likely leading to something not entirely desired, like all pods going to a single domain) or 
+focus on a single domain (thus not achieving the goal of being in two or more domains). 
+
+PodTopologySpreading solves the problem a bit, but not completely. It allows only even spreading 
+and once the deployment gets skewed it doesn’t do anything to rebalance. Pod topology spreading 
+(with skew and/or ScheduleAnyway flag) is also just a hint, if skewed placement is available and 
+allowed then Cluster Autoscaler is not triggered and the user ends up with a skewed deployment. 
+A user could specify a strict pod topolog spreading but then, in case of problems the deployment
+would not move its pods to the domains that are available. The growth of the deployment would also 
+be totally blocked as the available domains would be too much skewed.
+
+Thus, if full flexibility is needed, the only option is to have multiple deployments, targeting 
+different domains. This setup however creates one big problem. How to consistently autoscale multiple 
+deployments? The simplest idea - having multiple HPAs is not stable, due to different loads, race 
+conditions or so, some domains may grow while the others are shrunk. As HPAs and deployments are 
+not connected anyhow, the skewed setup will not fix itself automatically. It may eventually come to 
+a semi-balanced state but it is not guaranteed. 
+
+
+Thus there is a need for some component that will:
+
+* Keep multiple deployments aligned. For example it may keep an equal ratio between the number of
+pods in one deployment and the other. Or put everything to the first and overflow to the second and so on.
+* React to individual deployment problems should it be zone outage or lack of spot/preemptible vms. 
+* Actively try to rebalance and get to the desired layout.
+* Allow to autoscale all deployments with a single target, while maintaining the placement policy.
+
+## Balancer 
+
+Balancer is a stand-alone controller, living in userspace (or in control plane, if needed) exposing 
+a CRD API object, also called Balancer. Each balancer object has pointers to multiple deployments 
+or other pod-controlling objects that expose the Scale subresource. Balancer periodically checks 
+the number of running and problematic pods inside each of the targets, compares it with the desired 
+number of replicas, constraints and policies and adjusts the number of replicas on the targets, 
+should some of them run too many or too few of them. To allow being an HPA target Balancer itself
+exposes the Scale subresource.
+
```

Comment:
I don't know how much you want to go into details of how Balancer will work here and how much you want to decide later.

- When `Balancer.Spec.Replicas` changes balancer will resize controllers in `Balancer.Targets` so that their sizes sum to the new number of replicas and follow `Balancer.Spec.Policy`. I think this is clear.
- When a pod doesn't after `Balancer.Policy.Fallback.StartupTimeout` balancer assumes it won't start and instead increases a different Target. I have some questions:
  - Does it reduce size of the Target that didn't grow?
  - If it does then won't it cause problems?
  - If it does then is there a guarantee that the pod that didn't start will be removed (and not some other pod)? I think that Scale resource allows only to set size, not control which pods are removed.
- How does  Balancer decide its time to try to rebalance its Targets (after they don't have expected sizes because some pods didn't start)?
- Does balancer need to preserve any state in case it restarts?

Reply:
> Does it reduce size of the Target that didn't grow? 

No it doesn't. To keep the pressure on CA to eventually provide infrastructure. 

> How does Balancer decide its time to try to rebalance its Targets (after they don't have expected sizes because 
some pods didn't start)?

It reruns every 15 second and checks the sizes, pods, etc.

> Does balancer need to preserve any state in case it restarts?

Currently no.